irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -61,6 +62,82 @@ namespace {
 
 namespace iroha {
   namespace ametsuchi {
+
+    /**
+     * Class provides a reconnection for the session by manual manipilations
+     * Note: the class is a workaround for SOCI 4.0, supporting of further
+     * versions is not guaranteed
+     */
+    class FailOverCallback : public soci::failover_callback {
+     public:
+      FailOverCallback(
+          soci::postgresql_session_backend &connection,
+          std::shared_ptr<ReconnectionStrategy> reconnection_strategy,
+          logger::LoggerPtr log)
+          : connection_(connection_),
+            reconnection_strategy_(std::move(reconnection_strategy)),
+            log_(std::move(log)) {}
+
+      virtual void started() {
+        reconnection_strategy_->reset();
+        log_->debug("Reconnection process is initiated");
+      }
+
+      virtual void finished(soci::session &) {}
+
+      virtual void failed(bool &should_reconnect, std::string &) {
+        // don't rely on reconnection in soci because we are going to conduct
+        // our own reconnection process
+        should_reconnect = false;
+        log_->warn(
+            "troubles with storage connections are occurred. The system will "
+            "try to reconnect [{}]",
+            should_reconnect);
+        auto is_reconnected = reconnectionLoop();
+        log_->info("The connection re-established: {}", is_reconnected);
+      }
+
+      virtual void aborted() {
+        log_->error("has invoked aborted method of FailOverCallback");
+      }
+
+     private:
+      bool reconnectionLoop() {
+        bool successful_reconnection = false;
+        while (reconnection_strategy_->canReconnect()
+               and not successful_reconnection) {
+          try {
+            soci::connection_parameters parameters("postgresql", "");
+            connection_.clean_up();
+            connection_.connect(parameters);
+            successful_reconnection = true;
```

Comment:
Prepared statements and spdlog notice processor should be set here:
https://github.com/hyperledger/iroha/blob/5d481d4492bde61887c7692c64c38d2ff23f2eb6/irohad/ametsuchi/impl/storage_impl.cpp#L33
https://github.com/hyperledger/iroha/blob/5d481d4492bde61887c7692c64c38d2ff23f2eb6/irohad/ametsuchi/impl/storage_impl.cpp#L188